src/base/kaldi-error.cc

```diff
@@ -142,136 +146,76 @@ MessageLogger::MessageLogger(LogMessageEnvelope::Severity severity,
+               << envelope_.func  << "():" << envelope_.file << ':'
+               << envelope_.line << ") " << GetMessage().c_str();
+
+  // Add stack trace for errors and assertion failures, if available.
+  if (envelope_.severity < LogMessageEnvelope::kWarning) {
+    const std::string& stack_trace = KaldiGetStackTrace();
+    if (!stack_trace.empty()) {
+      full_message << "\n\n" << stack_trace;
     }
   }
-}
 
-FatalMessageLogger::FatalMessageLogger(LogMessageEnvelope::Severity severity,
-                                       const char *func, const char *file,
-                                       int32 line):
-  MessageLogger(severity, func, file, line) {
-  if (severity != LogMessageEnvelope::kAssertFailed &&
-      severity != LogMessageEnvelope::kError) {
-    // Don't use KALDI_ERR, since that will recursively instantiate
-    // MessageLogger.
-    throw std::runtime_error("FatalMessageLogger should be called only with "
-                             "severities kAssertFailed and kError");
-  }
-}
-#if defined(_MSC_VER)
-FatalMessageLogger::~FatalMessageLogger [[ noreturn ]] () noexcept(false)
-#else
-[[ noreturn ]] FatalMessageLogger::~FatalMessageLogger() noexcept(false)
-#endif
-{
-  std::string str = GetMessage();
-
-  // print the mesage (or send to logging handler),
-  MessageLogger::HandleMessage(envelope_, str.c_str());
-
-  // Should we throw exception, or abort?
-  switch (envelope_.severity) {
-    case LogMessageEnvelope::kAssertFailed:
-      abort(); // ASSERT_FAILED,
-      break;
-    case LogMessageEnvelope::kError:
-      if (!std::uncaught_exception()) {
-        // throw exception with empty message,
-        throw std::runtime_error(""); // KALDI_ERR,
-      } else {
-        // If we got here, this thread has already thrown exception,
-        // and this exception has not yet arrived to its 'catch' clause...
-        // Throwing a new exception would be unsafe!
-        // (can happen during 'stack unwinding', if we have 'KALDI_ERR << msg'
-        // in a destructor of some local object).
-        abort();
-      }
-      break;
-  default: // This should never happen, based on constructor's
-           // preconditions. But we place abort() here so that all
-           // possible pathways through this function do not return.
-    abort();
-  }
+  // Print the complete message to stderr.
+  full_message << "\n";
+  std::cerr << full_message.str();
 }
 
 
 /***** KALDI ASSERTS *****/
 
 void KaldiAssertFailure_(const char *func, const char *file,
                          int32 line, const char *cond_str) {
-  FatalMessageLogger ml(LogMessageEnvelope::kAssertFailed, func, file, line);
-  ml.stream() << ": '" << cond_str << "' ";
+  MessageLogger::Log() =
+    MessageLogger (LogMessageEnvelope::kAssertFailed, func, file, line)
+      << "Assertion failed: (" << cond_str << ")";
+  fflush(NULL);  // Flush all pending buffers, abort() may not flush stderr.
+  std::abort();
```

Comment:
within the c++ construct, is it better to rethrow a fatal exception vs a direct call to abort?

Reply:
This function only handles KALDI_ASSERT. It's reasonable to die at this point, because the macro says "it's impossible that the condition is false." You can throw around this abort in your logging function.

Reply:
ok, I see that.  Thanks